contracts/PrizeTierHistoryV2.sol

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.6;
+import "@pooltogether/owner-manager-contracts/contracts/Manageable.sol";
+import "./abstract/IdBinarySearch.sol";
+
+contract PrizeTierHistoryV2 is IdBinarySearch, Manageable {
+
+    struct PrizeTier {
+        uint8 bitRangeSize;
+        uint32 drawId;
+        uint32 maxPicksPerUser;
+        uint32 expiryDuration;
+        uint32 endTimestampOffset;
+        uint256 prize;
+        uint32[16] tiers;
+    }
+
+    uint32[] internal history;
+    mapping(uint32 => PrizeTier) internal prizeTiers;
+
+    /**
+     * @notice Emit when new PrizeTier is added to history
+     * @param drawId    Draw ID
+     * @param prizeTier PrizeTier parameters
+     */
+    event PrizeTierPushed(uint32 indexed drawId, PrizeTier prizeTier);
+
+    /**
+     * @notice Emit when existing PrizeTier is updated in history
+     * @param drawId    Draw ID
+     * @param prizeTier PrizeTier parameters
+     */
+    event PrizeTierSet(uint32 indexed drawId, PrizeTier prizeTier);
+
+    constructor(address owner, PrizeTier[] memory _history) Ownable(owner) {
+        if (_history.length > 0) {
+            inject(_history);
+        }
+    }
+
+    // @inheritdoc DrawIDBinarySearch
+    function getNewestIndex() internal view override returns (uint32) {
+        return uint32(history.length - 1);
+    }
+
+    // @inheritdoc DrawIDBinarySearch
+    function getIdForIndex(uint256 index) internal view override returns (uint32) {
+        return history[index];
+    }
+
+    function getPrizeTier(uint32 drawId) external view returns (PrizeTier memory) {
+        require(drawId > 0, "PrizeTierHistoryV2/draw-id-not-zero");
+        return prizeTiers[history[_binarySearch(drawId)]];
+    }
+
+    function getPrizeTierList(uint32[] calldata _drawIds) external view returns (PrizeTier[] memory) {
+        PrizeTier[] memory _data = new PrizeTier[](_drawIds.length);
+        for (uint256 index = 0; index < _drawIds.length; index++) {
+            _data[index] = prizeTiers[history[_binarySearch(_drawIds[index])]];
+        }
+        return _data;
+    }
+
+    function push(PrizeTier calldata nextPrizeTier) external onlyManagerOrOwner {
+        _push(nextPrizeTier);
+    }
+    
+    function popAndPush(PrizeTier calldata newPrizeTier) external onlyOwner {
+        uint length = history.length;
+        require(length > 0, "PrizeTierHistoryV2/history-empty");
+        require(history[length - 1] == newPrizeTier.drawId, "PrizeTierHistoryV2/invalid-draw-id");
+        _replace(newPrizeTier);
+    }
+
+    function replace(PrizeTier calldata newPrizeTier) external onlyOwner {
+        _replace(newPrizeTier);
+    }
+
+    function inject(PrizeTier[] memory timeline) public onlyOwner {
+        require(history.length == 0, "PrizeTierHistoryV2/history-not-empty");
+        require(timeline.length > 0, "PrizeTierHistoryV2/timeline-empty");
+        for (uint256 i = 0; i < timeline.length; i++) {
```

Comment:
Store `timeline.length` in a variable.

Reply:
I don't think we get any savings here with that since it's not loaded from storage. The pointer to timeline.length I would bet is actually cheaper then creating a new variable here.